jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -854,11 +854,23 @@ ecma_builtin_array_prototype_object_slice (ecma_value_t arg1, /**< start */
     {
       uint32_t length = end - start;
       ecma_extended_object_t *ext_to_obj_p = (ecma_extended_object_t *) new_array_p;
-      ecma_value_t *to_buffer_p = ecma_fast_array_extend (new_array_p, length);
+
+      ecma_value_t *to_buffer_p;
+      if (length == len)
+      {
+        to_buffer_p = ECMA_GET_NON_NULL_POINTER (ecma_value_t, new_array_p->u1.property_list_cp);
```

Comment:
@dbatyai @LaszloLango I have updated the PR, but on this line we could have a nullpointer which will makes a few tests fail. This patch will have to be rebased once #3480 lands, and then it should work.

Reply:
Yes, meanwhile I realized that my solution is not correct.
However it is not true that the target array can contain any number of elements. According to [standard](https://www.ecma-international.org/ecma-262/6.0/#sec-array.prototype.slice):
```
11. Let  count be max(final – k, 0).
12. Let A be ArraySpeciesCreate(O, count).
```
This way the target array is already created with the desired length, so if we follow the standard than there shouldn't be a need for either extending or shrinking in any case (with ES2015 profile). I think we should do it like this:
```c
#if ENABLED (JERRY_ES2015)
      ecma_value_t *to_buffer_p = ECMA_GET_NON_NULL_POINTER (ecma_value_t, new_array_p->u1.property_list_cp);
#else /* !ENABLED (JERRY_ES2015) */
      ecma_value_t *to_buffer_p = ecma_fast_array_extend (new_array_p, length);
#endif /* ENABLED (JERRY_ES2015) */
```
And of course the creation of `new_array` needs to be changed to follow the standard, because currently we create it with 0 length.
What do you think about this? I tried this locally and all tests passed with `tools/run-tests.py --jerry-tests`.

Reply:
You are correct, the constructor is called with incorrect arguments. However, even if the construct argument is fixed, the custom constructor can just ignore it, and create an arbitrary array. Try this:

```
class CustomArray extends Array {
    constructor(len) {
        super(42);
        this.fill ("str");
    }
}

var a = new CustomArray(5);  // Array length will be 42, regardless of the argument
a.length = 5;  // Crop the source array;
a.slice();  // Number of copied elements will be 5, but ArraySpeciesCreate will create an array with length 42
```

This is what I meant when I asked for more tests. We should add tests for when the constructor creates smaller/same sized/larger arrays than what is expected. I would also prefer a test for when the constructor creates a non-fast array, for example by adding a non array index property in the constructor.

Reply:
Good point, I did not consider the custom constructor.